File: src/main/java/utils/bytes/Xor.java
package utils.bytes;

public class Xor {
    public static byte[] sameLength(byte[] fst, byte[] snd) {
        if (fst.length != snd.length)
            throw new IllegalArgumentException("Input byte arrays must be of the same length");

        byte[] result = new byte[fst.length];

        for (int i = 0; i < result.length; i++) {
            result[i] = (byte) (fst[i] ^ snd[i]);
        }

        return result;
    }

    public static byte[] repeated(byte[] notRepeat, byte repeat) {
        return repeated(notRepeat, new byte[] { repeat });
    }

    public static byte[] repeated(byte[] notRepeat, byte[] repeat) {
        byte[] result = new byte[notRepeat.length];

        for (int i = 0; i < result.length; i++) {
            result[i] = (byte) (notRepeat[i] ^ repeat[i % repeat.length]);
        }

        return result;
    }
}

File: src/main/kotlin/utils/bytes/Xor.kt
package utils.bytes

@ExperimentalUnsignedTypes
object Xor {
    fun sameLength(fst: UByteArray, snd: UByteArray): UByteArray {
        require(fst.size == snd.size) { "Input byte arrays must be of the same length" }
        return fst.zip(snd).map { it.first xor it.second }.toUByteArray()
    }

    fun repeated(notRepeat: UByteArray, repeat: UByte): UByteArray {
        return repeated(notRepeat, ubyteArrayOf(repeat))
    }

    fun repeated(notRepeat: UByteArray, repeat: UByteArray): UByteArray {
        val result = UByteArray(notRepeat.size)
        for (i in result.indices) {
            result[i] = (notRepeat[i] xor repeat[i % repeat.size])
        }
        return result
    }
}
